api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -472,6 +480,18 @@ private Workspace updateWorkspaceImpl(
 
     dbWorkspace = workspaceDao.saveWithLastModified(dbWorkspace);
 
+    // Grant the workspace cloner and all from-workspaces users permission to use workflow if
+    // workspace is controlled tier workspace.
+    if (workbenchConfigProvider.get().featureFlags.grantLifescienceApiRunnerAcl
+        && accessTier.getShortName().equals(CONTROLLED_TIER_SHORT_NAME)) {
+      iamService.grantWorkflowRunnerRole(dbWorkspace.getGoogleProject());
+      for (Map.Entry<String, WorkspaceAccessLevel> entry : clonedRoles.entrySet()) {
+        if (shouldGrantWorkflowRunnerAsService(user, entry)) {
+          iamService.grantWorkflowRunnerRoleAsService(
```

Comment:
I'm concerned about:

1. consistency guarantees from these GCP APIs. I don't know whether it is strongly consistent, i.e. whether GET, UPDATE, GET, UPDATE, ... in rapid succession is guaranteed retrieve the up-to-date value (if not, this will drop some people from the ACL).
2. Slowness of applying this in serial; createWorkspace is already very slow.

I think it may make sense to create another method here which takes a list of users instead.

Reply:
Do you suggest to create another method that takes a list of users? But even have that, it does not resolve the two issues you mentioned. 
Get petSA and grant pet ActAs pet don't support batch. The only improvement is batch grant PAPI runner, and feels the improvement is small, also makes the code complicated(iterate over SAM getPetSA then batch papiRunner).
Two things I am considering:
(1) Most of workspaces does not have that many writers/owners? Or do we have some kind of limit on how many owners a workspace can have? If the average is lot(<5), I think it should be fine, since all new APIs I added is fast.
(2) Shall we make this a "best effort" async call? i.e. create a API `grantWorkflow` and separate from create/clone/share workspace call. So UI will fire two APIs, if the `grantWorkflow` fails, we tell user to retry(re-share).

Reply:
The first issue I mentioned is purely related to the GCP IAM APIs - my suggestion definitely solves that concern. It would also be faster to make 1 call instead of N calls to GCP for each of these cases, though you're right that it doesn't eliminate the serial getPetSA. I think the complexity is likely warranted, though maybe I'm missing something.

In general, I would say 3-5 collaborators is typical, but there are definitely some workspaces that have 10-12. I don't think we should split workflow roles from share/unshare. Moving the option of "share to existing collaborators" out of cloneWorkspace, and making it into two calls on the client (clone -> share) is another option. It's a bit sad we'd have to resort to this, but I would prefer to splitting out a different endpoint by role.

I guess we can see the performance of this and decide about splitting it. What is the latency you have observed so far?

Reply:
@calbach per offline discussion, now the workflow runner is in batch mode now. 
But to simply the logic and keep it clean, I still keep grantCurrentUser and grantUser separate.